### PR TITLE
Implement in and notIn in multiple visitors

### PR DESCRIPTION
**Pull request by jun-he**
Implement `in` and `notIn` in multiple visitors.

The project and the projectStrict in other transforms (e.g. bucket, dates, truncate) are implemented in https://github.com/apache/incubator-iceberg/pull/628.

### CODE REVIEW

**Comment by jun-he** [comment]
Rebase this PR on master.
@rdblue  can you help to review it? Thanks!

File: api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java [hunk 1]
```diff
@@ -269,11 +269,16 @@ public Boolean or(Boolean leftResult, Boolean rightResult) {
 
     @Override
     public <T> Boolean in(BoundReference<T> ref, Set<T> literalSet) {
-      return ROWS_MIGHT_MATCH;
+      // in(col, {X, Y}) => eq(col, x) OR eq(col, x)
+      if (literalSet.stream().anyMatch(v -> eq(ref, toLiteral(v)))) {
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java]: This should not convert each value to a literal on every invocation. Instead, this PR can add a `literals` method to `BoundSetPredicate` to access the values as literals. That way, the conversion happens once and can be lazy.

File: api/src/main/java/org/apache/iceberg/expressions/ExpressionVisitors.java [hunk 2]
```diff
@@ -59,6 +59,10 @@ public R or(R leftResult, R rightResult) {
   }
 
   public abstract static class BoundExpressionVisitor<R> extends ExpressionVisitor<R> {
+    protected <T> Literal<T> toLiteral(T val) {
+      return Literals.from(val);
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/expressions/ExpressionVisitors.java]: I don't see value in adding this method. Callers can use `Literals.from` directly, but the visitors should not even be converting values to literals.

File: api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java (hunk 1 above)
Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java]: It looks like this probably just needs the `Comparator`, not the literals.

File: api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java [hunk 3]
```diff
@@ -269,11 +269,16 @@ public Boolean or(Boolean leftResult, Boolean rightResult) {
 
     @Override
     public <T> Boolean in(BoundReference<T> ref, Set<T> literalSet) {
-      return ROWS_MIGHT_MATCH;
+      // in(col, {X, Y}) => eq(col, x) OR eq(col, x)
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java]: This should copy the logic from `eq` instead of calling it directly so that the lower and upper bounds are accessed and converted just once.

File: api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java [hunk 4]
```diff
@@ -269,11 +269,16 @@ public Boolean or(Boolean leftResult, Boolean rightResult) {
...
+      if (literalSet.stream().anyMatch(v -> eq(ref, toLiteral(v)))) {
+        return ROWS_MIGHT_MATCH;
+      }
+      return ROWS_CANNOT_MATCH;
     }
 
     @Override
     public <T> Boolean notIn(BoundReference<T> ref, Set<T> literalSet) {
+      // notIn(col, {X, Y}) => notEq(col, x) AND notEq(col, x)
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java]: This comment doesn't really make sense, since this doesn't delegate to `notEq`. Instead, can you copy and update the comment in `notEq`?

**Comment by rdblue** [comment]
Looks like this needs more tests updated. I think each updated eval visitor should have new tests.

@moulimukherjee, you might want to review this as well.

File: api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java (hunk 1 above)
Comment (jun-he) [reply api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java]: Thanks for the comments and update the code accordingly.

File: api/src/main/java/org/apache/iceberg/expressions/ExpressionVisitors.java (hunk 2 above)
Comment (jun-he) [reply api/src/main/java/org/apache/iceberg/expressions/ExpressionVisitors.java]: Removed.

File: api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java (hunk 3 above)
Comment (jun-he) [reply api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java]: Done.

File: api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java (hunk 4 above)
Comment (jun-he) [reply api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java]: Yep, updated.

**Comment by jun-he** [comment]
@rdblue Thanks for the code review. I have updated the PRs accordingly. 
Can you let me know if you have any additional comments? Thanks!

File: api/src/test/java/org/apache/iceberg/expressions/TestInclusiveManifestEvaluator.java [hunk 5]
```diff
@@ -347,4 +349,52 @@ public void testStringStartsWith() {
...
+
+  @Test
+  public void testIntegerIn() {
+    boolean shouldRead = ManifestEvaluator.forRowFilter(in("id", 5, 6), SPEC, true).eval(FILE);
+    Assert.assertFalse("Should not read: id below lower bound", shouldRead);
+
+    shouldRead = ManifestEvaluator.forRowFilter(in("id", 28, 29), SPEC, true).eval(FILE);
+    Assert.assertFalse("Should not read: id below lower bound", shouldRead);
+
+    shouldRead = ManifestEvaluator.forRowFilter(in("id", 30, 31), SPEC, true).eval(FILE);
```
Comment (moulimukherjee) [inline api/src/test/java/org/apache/iceberg/expressions/TestInclusiveManifestEvaluator.java]: nit: It might be more readable if the values 30 and 79 were extracted into min/max variables?

File: api/src/test/java/org/apache/iceberg/expressions/TestInclusiveMetricsEvaluator.java [hunk 6]
```diff
@@ -431,4 +433,52 @@ public void testStringStartsWith() {
...
+
+  @Test
+  public void testIntegerNotIn() {
+    boolean shouldRead = new InclusiveMetricsEvaluator(SCHEMA, notIn("id", 5, 6)).eval(FILE);
+    Assert.assertTrue("Should read: id below lower bound", shouldRead);
+
+    shouldRead = new InclusiveMetricsEvaluator(SCHEMA, notIn("id", 28, 29)).eval(FILE);
+    Assert.assertTrue("Should read: id below lower bound", shouldRead);
+
+    shouldRead = new InclusiveMetricsEvaluator(SCHEMA, notIn("id", 30, 31)).eval(FILE);
```
Comment (moulimukherjee) [inline api/src/test/java/org/apache/iceberg/expressions/TestInclusiveMetricsEvaluator.java]: Same as above. If the variables are extracted out, the tests can be in terms of min - 1, min + 1 etc.

**Comment by moulimukherjee** [comment]
Some nits, otherwise the changes look good to me.

File: api/src/test/java/org/apache/iceberg/expressions/TestInclusiveManifestEvaluator.java (hunk 5 above)
Comment (jun-he) [reply api/src/test/java/org/apache/iceberg/expressions/TestInclusiveManifestEvaluator.java]: Thanks for the comments. I think it is better to put additional info into `Assert` message.
In this way, it will also improve readability for other cases, e.g. values of `5, 6`.

File: api/src/test/java/org/apache/iceberg/expressions/TestInclusiveMetricsEvaluator.java (hunk 6 above)
Comment (jun-he) [reply api/src/test/java/org/apache/iceberg/expressions/TestInclusiveMetricsEvaluator.java]: Updated the `Assert` to improve the readability.

**Comment by jun-he** [comment]
@rdblue @moulimukherjee Updated the PR accordingly. Can you help to review?
Thanks!

File: api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java [hunk 7]
```diff
@@ -269,11 +270,38 @@ public Boolean or(Boolean leftResult, Boolean rightResult) {
 
     @Override
     public <T> Boolean in(BoundReference<T> ref, Set<T> literalSet) {
+      Integer id = ref.fieldId();
+
+      if (containsNullsOnly(id)) {
+        return ROWS_CANNOT_MATCH;
+      }
+
+      final Comparator<T> comparator = ((BoundSetPredicate<T>) expr).comparator();
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java]: This can't use `expr` because that is the expression being evaluated. There is no guarantee it is a bound set predicate.

Instead, I think this should add a `comparator()` method to `BoundReference` because it is passed into all of these methods. That has a `type()` accessor that you can use to get the comparator, like this:

```java
  Comparator<T> comparator() {
    return Comparators.forType(type().asPrimitiveType());
  }
```

File: api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java [hunk 8]
```diff
@@ -269,11 +270,38 @@ public Boolean or(Boolean leftResult, Boolean rightResult) {
...
+      Set<T> literals = literalSet;
+
+      if (lowerBounds != null && lowerBounds.containsKey(id)) {
+        T lower = Conversions.fromByteBuffer(ref.type(), lowerBounds.get(id));
+        literals = literals.stream().filter(v -> comparator.compare(lower, v) <= 0).collect(Collectors.toSet());
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java]: There's no need to use a set to collect the values, and the set would not correctly handle mixed UTF8 string types.

I think it would be better to use a list for the result. That would mean that `literals` can't be reused, but I think it is fine to use a separate `List<T>` variable for values that are greater than the lower bound.

File: api/src/main/java/org/apache/iceberg/expressions/ManifestEvaluator.java [hunk 9]
```diff
@@ -250,11 +251,33 @@ public Boolean or(Boolean leftResult, Boolean rightResult) {
 
     @Override
     public <T> Boolean in(BoundReference<T> ref, Set<T> literalSet) {
+      int pos = Accessors.toPosition(ref.accessor());
+      PartitionFieldSummary fieldStats = stats.get(pos);
+      if (fieldStats.lowerBound() == null) {
+        return ROWS_CANNOT_MATCH; // values are all null and literal cannot contain null
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/expressions/ManifestEvaluator.java]: Nit: literalSet cannot contain null.

File: api/src/main/java/org/apache/iceberg/expressions/ManifestEvaluator.java [hunk 10]
```diff
@@ -250,11 +251,33 @@ public Boolean or(Boolean leftResult, Boolean rightResult) {
...
+      }
+      final Comparator<T> comparator = ((BoundSetPredicate<T>) expr).comparator();
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/expressions/ManifestEvaluator.java]: No need to use `final`.

File: api/src/main/java/org/apache/iceberg/expressions/ManifestEvaluator.java [hunk 11]
```diff
@@ -250,11 +251,33 @@ public Boolean or(Boolean leftResult, Boolean rightResult) {
 
     @Override
     public <T> Boolean in(BoundReference<T> ref, Set<T> literalSet) {
+      int pos = Accessors.toPosition(ref.accessor());
+      PartitionFieldSummary fieldStats = stats.get(pos);
+      if (fieldStats.lowerBound() == null) {
+        return ROWS_CANNOT_MATCH; // values are all null and literal cannot contain null
+      }
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/expressions/ManifestEvaluator.java]: There should be a newline after control flow statements like this one.

File: api/src/main/java/org/apache/iceberg/expressions/StrictMetricsEvaluator.java [hunk 12]
```diff
@@ -308,11 +310,67 @@ public Boolean or(Boolean leftResult, Boolean rightResult) {
...
+
+      if (lowerBounds != null && lowerBounds.containsKey(id) &&
+          upperBounds != null && upperBounds.containsKey(id)) {
+        final Comparator<T> comparator = ((BoundSetPredicate<T>) expr).comparator();
+        Set<T> literals = literalSet;
+
+        T lower = Conversions.fromByteBuffer(struct.field(id).type(), lowerBounds.get(id));
+        literals = literals.stream().filter(v -> comparator.compare(lower, v) == 0).collect(Collectors.toSet());
+        if (literals.isEmpty()) {
+          return ROWS_MIGHT_NOT_MATCH;
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/expressions/StrictMetricsEvaluator.java]: I think this logic can be simplified to use the set. It should check if `lower` is in the set, check if `upper` is in the set, and finally check if `lower` and `upper` are equal. All values must be in the set if those three conditions are met. Can you update the logic here and add some comments to explain the implementation?

File: api/src/test/java/org/apache/iceberg/expressions/TestInclusiveManifestEvaluator.java [hunk 13]
```diff
@@ -347,4 +349,52 @@ public void testStringStartsWith() {
...
+
+  @Test
+  public void testIntegerIn() {
+    boolean shouldRead = ManifestEvaluator.forRowFilter(in("id", 5, 6), SPEC, true).eval(FILE);
+    Assert.assertFalse("Should not read: id below lower bound (5 < 30, 6 < 30)", shouldRead);
+
+    shouldRead = ManifestEvaluator.forRowFilter(in("id", 28, 29), SPEC, true).eval(FILE);
+    Assert.assertFalse("Should not read: id below lower bound (28 < 30, 29 < 30)", shouldRead);
+
+    shouldRead = ManifestEvaluator.forRowFilter(in("id", 30, 31), SPEC, true).eval(FILE);
```
Comment (rdblue) [inline api/src/test/java/org/apache/iceberg/expressions/TestInclusiveManifestEvaluator.java]: I think this test case should be `in("id", 29, 30)` to test the case where only one value is equal to or greater than the lower bound.

File: api/src/test/java/org/apache/iceberg/expressions/TestInclusiveMetricsEvaluator.java (hunk 6 above)
Comment (rdblue) [reply api/src/test/java/org/apache/iceberg/expressions/TestInclusiveMetricsEvaluator.java]: I agree with @moulimukherjee, it would be nice for the test cases to be in terms of file's min and max.

File: api/src/test/java/org/apache/iceberg/expressions/TestStrictMetricsEvaluator.java [hunk 14]
```diff
@@ -372,4 +374,47 @@ public void testIntegerNotEqRewritten() {
...
+
+    shouldRead = new StrictMetricsEvaluator(SCHEMA, in("id", 80, 81)).eval(FILE);
+    Assert.assertFalse("Should not match: some values != 80 and != 81)", shouldRead);
+
+    shouldRead = new StrictMetricsEvaluator(SCHEMA, in("always_5", 5, 6)).eval(FILE);
+    Assert.assertTrue("Should match: all values == 5", shouldRead);
+  }
+
+  @Test
+  public void testIntegernotIn() {
```
Comment (rdblue) [inline api/src/test/java/org/apache/iceberg/expressions/TestStrictMetricsEvaluator.java]: Typo: `testIntegerNotIn`.

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetDictionaryRowGroupFilter.java [hunk 15]
```diff
@@ -280,12 +281,33 @@ public Boolean or(Boolean leftResult, Boolean rightResult) {
...
-      return ROWS_MIGHT_MATCH;
+      int id = ref.fieldId();
+
+      Boolean hasNonDictPage = isFallback.get(id);
+      if (hasNonDictPage == null || hasNonDictPage) {
+        return ROWS_MIGHT_MATCH;
+      }
+
+      Set<T> dictionary = dict(id, ((BoundSetPredicate<T>) expr).comparator());
+      if (dictionary.size() > 1 || mayContainNulls.get(id)) {
```
Comment (rdblue) [inline parquet/src/main/java/org/apache/iceberg/parquet/ParquetDictionaryRowGroupFilter.java]: The case where `dictionary.size() > 1` only helps for a `notEq` predicate, like `!= "a"`. In that case, this is checking whether all values in the column are `"a"`. If there is more than one value, like `dict = Set("a", "b")`, then at least one value in the column must be the other dictionary entry and not all values are equal to the excluded one (so the row group should be read).

For a set of values, the equivalent logic is to check whether the dictionary size is greater than the number of literals in the `notIn` set. If so, then at least one value must not be in the set and the row group should be read.

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetDictionaryRowGroupFilter.java [hunk 16]
```diff
@@ -280,12 +281,33 @@ public Boolean or(Boolean leftResult, Boolean rightResult) {
...
+        return ROWS_MIGHT_MATCH;
+      }
+
+      return Sets.intersection(dictionary, literalSet).isEmpty() ? ROWS_MIGHT_MATCH : ROWS_CANNOT_MATCH;
```
Comment (rdblue) [inline parquet/src/main/java/org/apache/iceberg/parquet/ParquetDictionaryRowGroupFilter.java]: If any value in the dictionary is not in the set, then a row has a value that matches the `notIn` predicate. This tests whether all values of the dictionary are not in the set (the intersection is empty). That's not correct. For example, if the dictionary is `["a", "b", "c"]` and the predicate is `notIn("a", "b", "d")`, this would return `ROWS_CANNOT_MATCH` because the intersection is `Set("a", "b")`. But, at least one row has value `"c"` that is not in the set, so rows do match.

I think this should be a set difference operation. If `Sets.difference(dictionary, literalSet).isEmpty()` then there are no values in the dictionary that are not also in the set. All values are blacklisted so there are no values matching the predicate and this can return `ROWS_CANNOT_MATCH` in that case. Otherwise, rows might match.

File: parquet/src/test/java/org/apache/iceberg/parquet/TestDictionaryRowGroupFilter.java [hunk 17]
```diff
@@ -692,4 +694,66 @@ public void testMissingDictionaryPageForColumn() {
...
+    Assert.assertFalse("Should not read: id above upper bound (80 > 79, 81 > 79)", shouldRead);
+
+    shouldRead = new ParquetDictionaryRowGroupFilter(SCHEMA, in("id", 85, 86))
+        .shouldRead(parquetSchema, rowGroupMetadata, dictionaryStore);
+    Assert.assertFalse("Should not read: id above upper bound (85 > 79, 86 > 79)", shouldRead);
+  }
+
+  @Test
+  public void testIntegerNotIn() {
+    boolean shouldRead = new ParquetDictionaryRowGroupFilter(SCHEMA, notIn("id", 5, 6))
```
Comment (rdblue) [inline parquet/src/test/java/org/apache/iceberg/parquet/TestDictionaryRowGroupFilter.java]: All of these cases test the path where the size of the set is > 1.

This should test:
* The `notIn` set is a subset of the dictionary
* The dictionary is a subset of the `notIn` set
* The two sets are disjoint
* The two sets are equal

I'd also like to see some tests on the `some_nulls` column.

File: api/src/test/java/org/apache/iceberg/expressions/TestInclusiveManifestEvaluator.java [hunk 18]
```diff
@@ -347,4 +349,52 @@ public void testStringStartsWith() {
     shouldRead = ManifestEvaluator.forRowFilter(startsWith("some_nulls", "1"), SPEC, false).eval(FILE);
     Assert.assertFalse("Should skip: range doesn't match", shouldRead);
   }
+
+  @Test
+  public void testIntegerIn() {
```
Comment (rdblue) [inline api/src/test/java/org/apache/iceberg/expressions/TestInclusiveManifestEvaluator.java]: I think this also needs tests for all the null cases. Can you add tests for the `all_nulls` and `no_nulls` columns to all of the suites?

**Comment by rdblue** [comment]
Thanks @jun-he! Overall, this is looking close. I found a correctness problem with the Parquet dictionary implementation and I noted a few more test cases and cleanup.

Thanks to @moulimukherjee for reviewing as well!

File: api/src/test/java/org/apache/iceberg/expressions/TestInclusiveMetricsEvaluator.java (hunk 6 above)
Comment (jun-he) [reply api/src/test/java/org/apache/iceberg/expressions/TestInclusiveMetricsEvaluator.java]: 👌  will update all the tests.

File: api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java (hunk 7 above)
Comment (jun-he) [reply api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java]: Done.

File: api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java (hunk 8 above)
Comment (jun-he) [reply api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java]: Done.

File: api/src/main/java/org/apache/iceberg/expressions/ManifestEvaluator.java (hunk 9 above)
Comment (jun-he) [reply api/src/main/java/org/apache/iceberg/expressions/ManifestEvaluator.java]: Done.

File: api/src/main/java/org/apache/iceberg/expressions/ManifestEvaluator.java (hunk 11 above)
Comment (jun-he) [reply api/src/main/java/org/apache/iceberg/expressions/ManifestEvaluator.java]: Done.

File: api/src/main/java/org/apache/iceberg/expressions/StrictMetricsEvaluator.java (hunk 12 above)
Comment (jun-he) [reply api/src/main/java/org/apache/iceberg/expressions/StrictMetricsEvaluator.java]: Thanks for the comments. Updated the code and added comments accordingly.

File: api/src/test/java/org/apache/iceberg/expressions/TestInclusiveManifestEvaluator.java (hunk 13 above)
Comment (jun-he) [reply api/src/test/java/org/apache/iceberg/expressions/TestInclusiveManifestEvaluator.java]: Done.

File: api/src/test/java/org/apache/iceberg/expressions/TestStrictMetricsEvaluator.java (hunk 14 above)
Comment (jun-he) [reply api/src/test/java/org/apache/iceberg/expressions/TestStrictMetricsEvaluator.java]: Fixed.

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetDictionaryRowGroupFilter.java (hunk 15 above)
Comment (jun-he) [reply parquet/src/main/java/org/apache/iceberg/parquet/ParquetDictionaryRowGroupFilter.java]: Thanks for the comments. Fixed it.

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetDictionaryRowGroupFilter.java (hunk 16 above)
Comment (jun-he) [reply parquet/src/main/java/org/apache/iceberg/parquet/ParquetDictionaryRowGroupFilter.java]: Thanks for the comments! Fixed it.

File: parquet/src/test/java/org/apache/iceberg/parquet/TestDictionaryRowGroupFilter.java (hunk 17 above)
Comment (jun-he) [reply parquet/src/test/java/org/apache/iceberg/parquet/TestDictionaryRowGroupFilter.java]: Added more tests to cover them.

File: api/src/test/java/org/apache/iceberg/expressions/TestInclusiveManifestEvaluator.java (hunk 18 above)
Comment (jun-he) [reply api/src/test/java/org/apache/iceberg/expressions/TestInclusiveManifestEvaluator.java]: Done.

**Comment by jun-he** [comment]
Thanks @rdblue and @moulimukherjee for the code review!
Fixed the problems and added more tests.

**Comment by rdblue** [comment]
Looks like this has a conflict now. Could you fix it and I'll review again?

**Comment by jun-he** [comment]
@rdblue  Done. Can you please take a look? Thanks.

File: api/src/main/java/org/apache/iceberg/expressions/StrictMetricsEvaluator.java [hunk 19]
```diff
@@ -308,11 +310,69 @@ public Boolean or(Boolean leftResult, Boolean rightResult) {
...
+      Integer id = ref.fieldId();
+      Types.NestedField field = struct.field(id);
+      Preconditions.checkNotNull(field, "Cannot filter by nested column: %s", schema.findField(id));
+
+      if (canContainNulls(id)) {
+        return ROWS_MIGHT_NOT_MATCH;
+      }
+
+      if (lowerBounds != null && lowerBounds.containsKey(id) &&
+          upperBounds != null && upperBounds.containsKey(id)) {
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/expressions/StrictMetricsEvaluator.java]: Not something we need to fix, but I was thinking about this and wanted to note it: we may be able to do better than this for small ranges. For example, if the set is `Set(2, 3, 4, 5)` and the lower/upper bounds are `[2, 4]`, then the set does guarantee that all values are present. We could test this by detecting a small range and testing whether each value in the range is in the set.

**Comment by rdblue** [comment]
Merged! Thanks for getting this done, @jun-he!
